pkg/volume/metrics_du.go

```diff
@@ -0,0 +1,129 @@
+/*
+Copyright 2014 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package volume
+
+import (
+	"errors"
+	"fmt"
+	"os/exec"
+	"strconv"
+	"strings"
+)
+
+var _ VolumeMetricsProvider = &MetricsDu{}
+
+// MetricsDu represents a VolumeMetricsProvider that calculates the used and available
+// Volume capacity by executing the "du" and "df" commands on the Volume path.
+type MetricsDu struct {
+	// the directory path the volume is mounted to.
+	path string
+}
+
+// InitMetricsDu initializes MetricsDu with the Volume path.
+// This is required before MetricsDu can be used.
+func (cm *MetricsDu) InitMetricsDu(path string) {
+	cm.path = path
+}
+
+// See VolumeMetricsProvider.GetCapacityMetrics
+// GetCapacityMetrics calculates the volume usage and device free space by executing "du"
+// and "df" on path.
+func (cm *MetricsDu) GetCapacityMetrics() (*CapacityMetrics, error) {
+	metrics := &CapacityMetrics{}
+	if cm.path == "" {
+		return metrics, errors.New("no path defined for disk usage metrics.  Must call InitMetricsDu.")
+	}
+
+	err := cm.runDu(metrics)
+	if err != nil {
+		return metrics, err
+	}
+
+	err = cm.runDf(metrics)
+	if err != nil {
+		return metrics, err
+	}
+
+	return metrics, nil
+}
+
+// runDu executes the "du" command and writes the results to metrics.VolumeBytesUsed
+func (cm *MetricsDu) runDu(metrics *CapacityMetrics) error {
+	// Uses the same niceness level as cadvisor.fs does when running du
+	// Uses -B 1 to always scale to a blocksize of 1 byte
+	out, err := exec.Command("nice", "-n", "19", "du", "-s", "-B", "1", cm.path).CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("failed command 'du' on %s with error %v", cm.path, err)
+	}
+	bUsed, err := strconv.ParseUint(strings.Fields(string(out))[0], 10, 64)
+	if err != nil {
+		return fmt.Errorf("failed to parse 'du' output %s due to error %v", out, err)
+	}
+	metrics.VolumeBytesUsed = bUsed
+	return nil
+}
+
+// runDf executes the "df" command and writes the results to metrics.DeviceBytesUsed
+// and metrics.DeviceBytesAvailable
+// TODO(pwittrock): Consider using a statfs syscall directly to get the data
```

Comment:
I'd prefer switching to statfs right away. For some volumes statfs can be used to get both capacity and usage. So the statfs logic should be abstracted appropriately.